---
title: "<center><div class='mytitle'>Differential Gene Expression (DGE) Analysis Report</div></center>"
date: "<center><div class='mydate'>`r Sys.Date()`</div></center>"
output:
  html_document:
      css: style.css
      toc: FALSE
      includes: 
        before_body: header.html
        after_body: footer.html
params:
  comparison_name: HDM10_s_vs_HDM10_s_MSC_l
  dge_results_in: top-up-mixed-asthma-DGE_results_MICE-MT_MT-REMOD.xlsx
  metadata: MT-REMOD_MICE-MT_Metadata.xlsx
  norm_counts: normalized_counts_no-miRNAs_MICE-MT_MT-REMOD.xlsx
  raw_counts: raw_counts_no-miRNAs_MICE-MT_MT-REMOD.xlsx
  organism: mus_musculus
---

```{r 0_setup, include=FALSE}
# Import font awesome for proper footer
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
```

```{r setup, include=FALSE}
## Global options
knitr::opts_chunk$set(
  cache = FALSE,
  echo = TRUE,
  message = FALSE,
  warning = FALSE
)
```

```{r experiment_setup, include=FALSE}
## Markdown Analysis Settings
`%nin%` = Negate(`%in%`) # negate %in%
## Dependencies
library(readxl)
library(tidyverse)
library(DESeq2)
library(ggthemes)
library(ggplot2)
library(plotly)
library(dendextend)
library(heatmaply)

 if (params$organism == 'homo_sapiens') {
  organism_keg <- 'hsa'
  organism_wikipath <- 'Homo sapiens'
  org_db <- org.Hs.eg.db::org.Hs.eg.db
  biomaRt_annotation_table <- biomaRt_annotation_homo_sapiens
} else if (params$organism == 'mus_musculus') {
  organism_keg <- 'mmu'
  organism_wikipath <- 'Mus musculus'
  org_db <- org.Mm.eg.db::org.Mm.eg.db
  biomaRt_annotation_table <- biomaRt_annotation_mus_musculus
} else {
  stop(message('Organism not supported...'))
}

padj_cutoff <- 0.05

log2foldChange_cutoff <- 1

comparison_name <- params$comparison_name

dge_results_in <- params$dge_results_in %>% 
  tibble::rownames_to_column(var = 'Geneid')

compared_samples <- dge_results_in$compared_samples[!is.na(dge_results_in$compared_samples)]

metadata <- params$metadata %>% 
  filter(sample %in% compared_samples) %>% 
  mutate(group = condition)

norm_counts <- params$norm_counts %>% 
  select(Geneid, compared_samples)

raw_counts <- params$raw_counts %>% 
  select(Geneid, compared_samples)

```

## Analysis Settings

**`r comparison_name`**

**padj threshold: `r padj_cutoff`**

**Log2FC threshold: `r log2foldChange_cutoff`**


# `r comparison_name` {.tabset}

## Overview

The following shows the results of the DGE analyses of 68 samples in total (samples MICE4 and MICE38 have been removed due to low QC statistics). Table below contains all of the sample metadata in this study.


```{r, echo = FALSE}
DT::datatable(metadata, rownames = FALSE, filter="top", options = list(pageLength = 1000, scrollX=F, columnDefs = list(list(className = 'dt-center', targets = 0:(ncol(metadata)-1)))))
```


## PCA

```{r calculations_pca2d, echo=FALSE}
# PCA on DESeq2 normalized counts -----------------------------------------
normalized_counts <- norm_counts %>% 
  column_to_rownames(var = 'Geneid')

# Perform PCA focusing only on first two dimensions for 2D plot
normalized_pca_2d <- normalized_counts %>%
  t() %>%
  prcomp(rank. = 2)

# Call summary on the PCA to acquire explained variance
normalized_pca_2d_sum <-
  summary(normalized_pca_2d)

# Create label for PCA showing % of explained variance
normalized_pca_2d_sum_total <-
  sum(normalized_pca_2d_sum$importance[2, 1:2]) * 100

```

##### 2D PCA on DESeq2's Normalized Counts

##### Total Explained Variance = `r normalized_pca_2d_sum_total`%

```{r plot_normalized_pca2d_all_i, echo=FALSE, fig.height=4, fig.width=7.5}

# create info for plot
pca2d_info <- data.frame(sample = rownames(normalized_pca_2d$x)) %>% 
  left_join(metadata)

Condition <- pca2d_info$group
Sample <- pca2d_info$sample

if (sum(pca2d_info$sample == metadata$sample) == length(metadata$sample) &
    sum(rownames(normalized_pca_2d$x) == Sample) == length(rownames(normalized_pca_2d$x))) {
  # Create PCA plot
normalized_pca2d_plot_all <- ggplot(data.frame(normalized_pca_2d$x), 
                                    aes(PC1, 
                                        PC2, 
                                        color = Condition, 
                                        label = Sample)) +
  geom_jitter(size = 3) +
  theme_hc()

# Convert PCA plot to be interactive 
(normalized_pca2d_plot_all_i <- normalized_pca2d_plot_all %>% 
    ggplotly() %>% 
    layout(margin=list(t = 75)))
} else {
  stop('Order of samples in PCA dataframe does not match the order of sampels in metadata (groups not proper annotated)')
}



```

```{r calculations_pca3d, echo = FALSE}
# Perform PCA focusing only on first two dimensions for 3D plot
normalized_pca_3d <- normalized_counts %>%
  t() %>%
  prcomp(rank. = 3)

# Call summary on the PCA to acquire explained variance
normalized_pca_3d_sum <-
  summary(normalized_pca_3d)

# Create label for PCA showing % of explained variance
normalized_pca_3d_sum_total <-
  sum(normalized_pca_3d_sum$importance[2, 1:3]) * 100

```

##### 3D PCA on DESeq2's Normalized Counts

##### Total Explained Variance = `r normalized_pca_3d_sum_total`%

```{r plot_normalized_pca3d_all_i, echo=FALSE, fig.height=4, fig.width=7.5}

# create info for plot
pca3d_info_3d <- data.frame(sample = rownames(normalized_pca_3d$x)) %>% 
  left_join(metadata)

Condition_3d <- pca3d_info_3d$group
Sample_3d <- pca3d_info_3d$sample

if (sum(pca3d_info_3d$sample == metadata$sample) == length(metadata$sample) &
    sum(rownames(normalized_pca_3d$x) == Sample) == length(rownames(normalized_pca_3d$x))) {
  # Create PCA plot
normalized_pca3d_plot_all_i <-
  plot_ly(
    data.frame(normalized_pca_3d$x),
    text = ~ rownames(normalized_pca_3d$x),
    x = ~ PC1,
    y = ~ PC2,
    z = ~ PC3,
    color = ~ Condition_3d,
    colors = c('#BF382A', '#0C4B8E')
  ) %>%
  add_markers() %>%
  layout(scene = list(
    xaxis = list(title = paste(
      "PC1 -",
      round(normalized_pca_3d_sum$importance[2, 1], 3)
    )),
    yaxis = list(title = paste(
      "PC2 -",
      round(normalized_pca_3d_sum$importance[2, 2], 3)
    )),
    zaxis = list(title = paste(
      "PC3 -",
      round(normalized_pca_3d_sum$importance[2, 3], 3)
    ))
  ))


normalized_pca3d_plot_all_i

} else {
  stop('Order of samples in 3D PCA dataframe does not match the order of sampels in metadata (groups not proper annotated)')
}


```

```{r calculations_pca2d_degs, echo = F}
# PCA on DESeq2 normalized counts - ONLY DEGs -----------------------------
# https://plotly.com/r/3d-scatter-plots/
# Filter DE results for DEGs
degs <- dge_results_in %>% 
  filter(padj < padj_cutoff) %>% 
  filter(abs(log2FoldChange) > log2foldChange_cutoff)

# Filter normalized matrix for DEGs
normalized_counts_degs <- normalized_counts[rownames(normalized_counts) %in% degs$Geneid,]

# Perform 2D PCA using only DE genes on a 
normalized_pca_2d_degs <- normalized_counts_degs %>%
  t() %>%
  prcomp(rank. = 2)

# Call summary on the PCA to acquire explained variance
normalized_pca_2d_degs_sum <- summary(normalized_pca_2d_degs)

# Create label for PCA showing % of explained variance
tot_explained_variance_ratio_2d_degs <- sum(normalized_pca_2d_degs_sum$importance[2,1:2]) * 100

if (sum(rownames(normalized_pca_2d_degs$x) == Sample) == length(rownames(normalized_pca_2d_degs$x))) {
# Create 2D PCA plot
normalized_pca2d_plot_degs <- ggplot(data.frame(normalized_pca_2d_degs$x), 
                                     aes(PC1, 
                                         PC2, 
                                         color = Condition, 
                                         label = Sample)) +
  geom_jitter(size = 3) +
  theme_hc()

# Convert 2D PCA plot into interactive using plotly
normalized_pca2d_plot_degs_i <- normalized_pca2d_plot_degs %>% 
  ggplotly() %>% 
  layout(margin=list(t = 75))
} else {
  stop('Order of samples in 3D PCA dataframe does not match the order of sampels in metadata (groups not proper annotated)')
}

```

##### 2D PCA on DESeq2's Normalized Counts - Only DEGs

##### Total Explained Variance = `r tot_explained_variance_ratio_2d_degs`% 

```{r pca2d_2, echo = F, fig.height = 4, fig.width=7.5}
normalized_pca2d_plot_degs_i
```

```{r calculations_pca3d_degs, echo = FALSE}
# Perform PCA focusing only on first three dimensions
normalized_pca_3d_degs <- normalized_counts[rownames(normalized_counts) %in% degs$Geneid,] %>%
  t() %>%
  prcomp(rank. = 3)

# Call summary on the PCA to acquire explained variance
normalized_pca_3d_degs_sum <-
  summary(normalized_pca_3d_degs)

# Create label for PCA showing % of explained variance
tot_explained_variance_ratio_3d_degs <-
  sum(normalized_pca_3d_degs_sum$importance[2, 1:3]) * 100

```

##### 3D PCA on DESeq2's Normalized Counts - Only DEGs

##### Total Explained Variance = `r tot_explained_variance_ratio_3d_degs`%

```{r plot_normalized_pca3d_plot_degs, echo = FALSE, fig.height = 4, fig.width=7.5}
if (sum(rownames(normalized_pca_3d_degs$x) == Sample_3d) == length(rownames(normalized_pca_3d_degs$x))) {
(normalized_pca3d_plot_degs <-
    plot_ly(
      data.frame(normalized_pca_3d_degs$x),
      text = ~rownames(normalized_pca_3d_degs$x),
      x = ~ PC1,
      y = ~ PC2,
      z = ~ PC3,
      color = ~ Condition_3d,
      colors = c('#BF382A', '#0C4B8E')
    ) %>%
    add_markers() %>%
    layout(scene = list(
      xaxis = list(title = paste("PC1 -", round(normalized_pca_3d_degs_sum$importance[2, 1], 3))),
      yaxis = list(title = paste("PC2 -", round(normalized_pca_3d_degs_sum$importance[2, 2], 3))),
      zaxis = list(title = paste("PC3 -", round(normalized_pca_3d_degs_sum$importance[2, 3], 3)))
    )))
} else {
  stop('Order of samples in 3D PCA dataframe does not match the order of sampels in metadata (groups not proper annotated)')
}

```

## Dendrograms

##### Dendrogram on DESeq2's Normalized Counts

```{r dendrogram_all, echo = FALSE, fig.height = 7, fig.width = 6.5}
# Dendrogram on DESeq2 normalized counts  ---------------------------------
# Create distance matrix
normalized_counts_all_d <- dist(t(normalized_counts))
normalized_counts_all_d_h <- hclust(normalized_counts_all_d, method = "complete")
levelsCondition <- factor(c(unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[6], unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[8])) 

dend_all <- as.dendrogram(normalized_counts_all_d_h)

# Color the branches based on the clusters:
dend_all_c <- color_branches(dend_all, k = 1, col = c('black'))#, groupLabels = levelsCondition)

# Manually match the labels, as much as possible, to the real classification of the flowers:
labels_colors(dend_all_c) <-
  c('#BF382A', '#0C4B8E')[sort_levels_values(as.numeric(Condition)[order.dendrogram(dend_all_c)])]

# We shall add the flower type to the labels:
labels(dend_all_c) <-
  paste(as.character(Condition)[order.dendrogram(dend_all_c)],
        "(", labels(dend_all_c), ")",
        sep = "")

# We hang the dendrogram a bit:
dend_all_c_h <- hang.dendrogram(dend_all_c, hang_height = 0.1)

# reduce the size of the labels:
dend_all_c_h_s <- set(dend_all_c_h, "labels_cex", 0.6)

# And plot:
par(mar = c(2, 2, 2, 14) + 0.1, xpd = NA)

{plot(dend_all_c_h_s,
     horiz =  TRUE,
     nodePar = list(cex = .007))

legend("topleft",
       legend = levelsCondition,
       fill = c('#0C4B8E', '#BF382A'))

dendextend::colored_bars(
  colors = ifelse(as.character(Condition) == paste0(unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[8]), '#BF382A', '#0C4B8E'),
  dend = dend_all_c_h_s,
  rowLabels = "",
  add = TRUE,
  horiz = TRUE
)}

```

##### Dendrogram on DESeq2 normalized counts with only DEGs

```{r dendrogram_degs, echo = FALSE, fig.height = 7, fig.width = 6.5}
# Dendrogram on DESeq2 normalized counts | DEGs only ---------------------
normalized_counts_degs <- normalized_counts[rownames(normalized_counts) %in% degs$Geneid,]

# Create distance matrix
normalized_counts_degs_d <- dist(t(normalized_counts_degs))
normalized_counts_degs_d_h <- hclust(normalized_counts_degs_d, method = "complete")
levelsCondition <- factor(c(unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[6], unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[8]))

dend_degs <- as.dendrogram(normalized_counts_degs_d_h)

# Color the branches based on the clusters:
dend_degs_c <- color_branches(dend_degs, k = 1, col = c('black'))#, groupLabels = levelsCondition)

# Manually match the labels, as much as possible, to the real classification of the flowers:
labels_colors(dend_degs_c) <-
  c('#0C4B8E','#BF382A')[sort_levels_values(as.numeric(Condition)[order.dendrogram(dend_degs_c)])]

# We shall add the flower type to the labels:
labels(dend_degs_c) <-
  paste(as.character(Condition)[order.dendrogram(dend_degs_c)],
        "(", labels(dend_degs_c), ")",
        sep = "")

# We hang the dendrogram a bit:
dend_degs_c_h <- hang.dendrogram(dend_degs_c, hang_height = 0.1)

# reduce the size of the labels:
dend_degs_c_h_s <- set(dend_degs_c_h, "labels_cex", 0.6)

# And plot:
par(mar = c(2, 2, 2, 14) + 0.1, xpd = NA)

{plot(dend_degs_c_h_s,
     horiz =  TRUE,
     nodePar = list(cex = .007))

legend("topleft",
       legend = levelsCondition,
       fill = c('#0C4B8E', '#BF382A'))

colored_bars(
  colors = ifelse(as.character(Condition) == paste0(unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[8]), '#BF382A', '#0C4B8E'),
  dend = dend_degs_c_h_s,
  rowLabels = "",
  add = TRUE,
  horiz = TRUE
)}

```


## Correlation Matrix

```{r calculations_heatmap_distance, echo = FALSE}
# Heatmap distance --------------------------------------------------------
# Heatmaply's gradient function is broken hence it will always use max and min matrix values as limits | ComplexHeatmap shiny app is better but as its an app markdown needs to be hosted in order to work
gradient_col <- ggplot2::scale_fill_gradient2(
  low = "blue", high = "red", 
  midpoint = 0, limits = c(-1, 1)
)

meta <- metadata %>% 
  mutate(condition = group) %>% 
  mutate(samples = sample)

# Input matrix
normalized_counts_d_r <- dist(t(normalized_counts),
                              diag = TRUE,
                              upper = TRUE) %>% 
  as.matrix() %>% 
  cor()

# Colnames and rownames pasted with its condition
colnames(normalized_counts_d_r) <- paste(meta$condition, meta$samples, sep = "-")
rownames(normalized_counts_d_r) <- colnames(normalized_counts_d_r)
sorted_normalized_counts_d_r <- normalized_counts_d_r[order(rownames(normalized_counts_d_r)), order(colnames(normalized_counts_d_r), decreasing = FALSE)]

# Relative height and width to the number of samples
normalized_counts_d_r_h = nrow(sorted_normalized_counts_d_r) * 0.45
normalized_counts_d_r_w = ncol(sorted_normalized_counts_d_r) * 0.55
```

```{r plot_heatmap_distance, echo = F, fig.height = 8, fig.width = 8.5}

# Side annotations
side_annot <- sub("-.*", "", colnames(sorted_normalized_counts_d_r))

side_annot_top <- sub("-.*", "", rownames(sorted_normalized_counts_d_r))

# Construct the heatmap
normalized_counts_d_r_heatmap <- heatmaply(
  sorted_normalized_counts_d_r,
  scale_fill_gradient_fun = gradient_col,
  col_side_colors = data.frame("Condition" = side_annot_top),
  row_side_colors = data.frame("Condition" = side_annot),
  limits = c(-1,1),
  Rowv = FALSE,
  Colv = FALSE,
  showticklabels = TRUE,
  row_side_palette = colorRampPalette(c('#0C4B8E', '#BF382A'), bias=1),
  col_side_palette = colorRampPalette(c('#0C4B8E', '#BF382A'), bias=1),
  key.title = 'corr',
  colorbar_len = 0.01,
  side_color_colorbar_len = 0.01,
  subplot_heights = c(0.05,0.95),
  subplot_widths = c(0.95,0.05),
  fontsize_row = 7,
  fontsize_col = 7) 

# Remove the duplicated labels caused by two sided annotations and remove the label
s <- subplot(normalized_counts_d_r_heatmap, margin = .01, titleY = TRUE) %>% 
  config(displayModeBar = FALSE)
s$x$data[[1]]$showlegend <- FALSE
s$x$data[[11]]$showlegend <- FALSE
#s$x$layout$annotations[[2]]$text <- ''
#s$x$layout$annotations[[2]]$y <- 0.9
#s$x$layout$annotations[[2]]$x <- 1.026

s

```

## Volcano Plots

```{r volcanos, echo = FALSE, fig.height = 12, fig.width=20}
# Volcano -----------------------------------------------------------------
# Set color and labels according to Log2FC and padj

# hgnc_ids <- biomaRt::getBM(
#     attributes = c('ensembl_gene_id', 'external_gene_name'),
#     filters = 'ensembl_gene_id',
#     values =  dge_results_in$Geneid,
#     mart = biomaRt::useDataset(
#         "mmusculus_gene_ensembl",
#         mart = biomaRt::useMart("ensembl", host = "asia.ensembl.org")
#     )
# ) %>%
#     dplyr::rename('Geneid' = 'ensembl_gene_id',
#                   'hgnc' = 'external_gene_name') %>%
#   full_join(dge_results_in)

hgnc_ids <- dge_results_in %>% left_join(biomaRt_annotation_table %>% 
  dplyr::select(ensembl_gene_id, external_gene_name) %>% 
  dplyr::distinct(ensembl_gene_id, .keep_all = TRUE) %>% 
    dplyr::rename('Geneid' = 'ensembl_gene_id',
                  'hgnc' = 'external_gene_name'))


keyvals <-
  ifelse(dge_results_in$Geneid %in% c(""),
         'red',
         ifelse((
           dge_results_in$log2FoldChange > as.numeric(log2foldChange_cutoff) |
             dge_results_in$log2FoldChange < -as.numeric(log2foldChange_cutoff)
         ) & dge_results_in$padj < as.numeric(padj_cutoff),
         'orange',
         ifelse(dge_results_in$padj > as.numeric(padj_cutoff), '#143D59', '#143D59')
         ))

keyvals[is.na(keyvals)] <- '#143D59'
#names(keyvals)[keyvals == 'red'] <- 'Commonly Regulated'
names(keyvals)[keyvals == '#143D59'] <-
  'Not Statistically Significant'
names(keyvals)[keyvals == 'orange'] <- 'Log2FC & padj'

# Construct volcano with labels

v1 <- EnhancedVolcano::EnhancedVolcano(
  dge_results_in,
  lab = hgnc_ids$hgnc,
  x = 'log2FoldChange',
  y = 'padj',
  title = paste0(unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[6], '_vs_', unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[8]),
  pCutoff = padj_cutoff,
  FCcutoff = log2foldChange_cutoff,
  pointSize = 3.5,
  labSize = 4.5,
  colCustom = keyvals,
  #col=c('#143D59', '#143D59', '#143D59', 'orange'),
  labCol = 'black',
  labFace = 'bold',
  boxedLabels = T,
  colAlpha = 3.5 / 5,
  legendPosition = 'top',
  legendLabSize = 18,
  legendIconSize = 5.0,
  drawConnectors = T,
  widthConnectors = 1.5,
  colConnectors = 'black',
  gridlines.major = T,
  gridlines.minor = FALSE,
  border = 'full',
  borderWidth = 1,
  borderColour = 'black',
  #xlim = c(-4,4),
  #ylim = c(0,11)
)

# Construct volcano without labels
v2 <- EnhancedVolcano::EnhancedVolcano(
  dge_results_in,
  lab = hgnc_ids$hgnc,
  x = 'log2FoldChange',
  y = 'padj',
  title = paste0(unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[6], '_vs_', unlist(stringr::str_split(dge_results_in$analysis_description[2], ' '))[8]),
  pCutoff = padj_cutoff,
  FCcutoff = log2foldChange_cutoff,
  pointSize = 3.5,
  labSize = 4.5,
  colCustom = keyvals,
  #col=c('#143D59', '#143D59', '#143D59', 'orange'),
  labCol = 'black',
  labFace = 'bold',
  boxedLabels = T,
  colAlpha = 3.5 / 5,
  legendPosition = 'top',
  legendLabSize = 18,
  legendIconSize = 5.0,
  drawConnectors = T,
  widthConnectors = 1.5,
  colConnectors = 'black',
  gridlines.major = T,
  gridlines.minor = FALSE,
  border = 'full',
  borderWidth = 1,
  borderColour = 'black',
  #xlim = c(-4,4),
  #ylim = c(0,11),
  selectLab = c("")
)

# Plot both
library(gridExtra)
grid.arrange(v1, v2, ncol=2)

```

```{r ggdc_plotly, echo = F, fig.height = 12, fig.width = 8.8}
# Data processing for GSEA Algorithm to GO and KEGG
library(clusterProfiler)

# Convert ENSEMBLID to ENTREZ
dge_entrez <- bitr(
  dge_results_in$Geneid,
  fromType = "ENSEMBL",
  toType = c("ENTREZID"),
  OrgDb = org_db
) %>%
  dplyr::rename(entrez = ENTREZID,
                Geneid = ENSEMBL) %>%
  full_join(dge_results_in)

# Create geneList from the whole data set for GSEA
geneList_dge <- dge_entrez$log2FoldChange
names(geneList_dge) <- dge_entrez$entrez
geneList_dge <- sort(geneList_dge, decreasing = TRUE)
geneList_dge_no_dups <-
  geneList_dge[!duplicated(names(geneList_dge))]

# Also create gene list with genes passing the Log2FC
gene <- names(geneList_dge_no_dups)[abs(geneList_dge_no_dups) > log2foldChange_cutoff]

# Also create gene list for DE genes
dge_entrez_ora <- dge_entrez %>%
  filter(padj < padj_cutoff)
geneList_dge_ora <- dge_entrez_ora$log2FoldChange
names(geneList_dge_ora) <- dge_entrez_ora$entrez
geneList_dge_ora <- sort(geneList_dge_ora, decreasing = TRUE)
geneList_dge_no_dups_ora <-
  geneList_dge_ora[!duplicated(names(geneList_dge_ora))]
gene_degs <-
  names(geneList_dge_no_dups_ora)[abs(geneList_dge_no_dups_ora) > log2foldChange_cutoff]
gene_degs_fc <-
  geneList_dge_no_dups_ora[abs(geneList_dge_no_dups_ora) > log2foldChange_cutoff]
```

```{r GSEA_GO, echo = F, fig.height = 12, fig.width = 8.8}
# After removing duplicates set seed for reproducibility
the_seed = 1623250700 %% 100000 # 50700

# set the same seed for the following analysis using the GSEA algorithm
# GENE ONTOLOGY
set.seed(the_seed)
go_all_gsea_dge_out <-
  clusterProfiler::gseGO(
    geneList     = geneList_dge_no_dups,
    OrgDb        = org_db,
    ont          = "all",
    pvalueCutoff = 1,
    verbose      = TRUE,
    seed         = TRUE,
    minGSSize    = 15,
    maxGSSize    = 500,
    keyType = 'ENTREZID'
  )
```

```{r GSEA_KEGG, echo = F, fig.height = 12, fig.width = 8.8}
# KYOTO ENCYCLOPEDIA OF GENES AND GENOMES
set.seed(the_seed)
kegg_gsea_dge_out <- gseKEGG(
  geneList     = geneList_dge_no_dups,
  organism     = organism_keg,
  minGSSize    = 15,
  maxGSSize    = 500,
  pvalueCutoff = 1,
  verbose      = TRUE,
  seed         = TRUE,
  use_internal_data = FALSE # seems deprecated hence has to be used online
)
```

```{r GSEA_rest, echo = F, fig.height = 12, fig.width = 8.8}
# WIKI PATHWAYS
set.seed(the_seed)
wiki_pathways_gsea <-
  gseWP(geneList_dge_no_dups, organism = organism_wikipath)
# 
# # REACTOME
# library(DOSE)
# library(ReactomePA)
# set.seed(the_seed)
# reactome_gsea <- gsePathway(
#   geneList_dge_no_dups,
#   pvalueCutoff = 1,
#   pAdjustMethod = "BH",
#   verbose = TRUE
# )
# 
# # DISEASE ONTOLOGY - disease ontology
# set.seed(the_seed)
# disease_ontology_do <- gseDO(
#   geneList_dge_no_dups,
#   minGSSize     = 15,
#   pvalueCutoff  = 1,
#   pAdjustMethod = "BH",
#   verbose       = TRUE
# )
# 
# # DISEASE ONTOLOGY - network of cancer gene
# set.seed(the_seed)
# disease_ontology_ncg <- gseNCG(
#   geneList_dge_no_dups,
#   minGSSize     = 15,
#   pvalueCutoff  = 1,
#   pAdjustMethod = "BH",
#   verbose       = TRUE
# )
# 
# # DISEASE ONTOLOGY - disease gene network
# set.seed(the_seed)
# disease_ontology_dgn <- gseDGN(
#   geneList_dge_no_dups,
#   minGSSize     = 15,
#   pvalueCutoff  = 1,
#   pAdjustMethod = "BH",
#   verbose       = TRUE
# )

```


## GO
### Top 30 - Gene Ontology

```{r print_GO, echo=F}
#DT::datatable((data.frame(go_all_gsea_dge_out))[1:30,])
DT::datatable((data.frame(go_all_gsea_dge_out)))
```

## KEGG
### Top 30 - Kyoto Encyclopedia of Genes and Genomes

```{r print_KEGG, echo=F}
#DT::datatable((data.frame(kegg_gsea_dge_out))[1:30,])
DT::datatable((data.frame(kegg_gsea_dge_out)))
```

## WIKI Pathways
### Top 30

```{r print_WIKIp, echo=F}
#DT::datatable((data.frame(wiki_pathways_gsea))[1:nrow(data.frame(wiki_pathways_gsea)),])
DT::datatable((data.frame(wiki_pathways_gsea)))
```

<!-- ## Reactome Pathways -->
<!-- ### Top 30 -->

<!-- ```{r print_reactome, echo=F} -->
<!-- DT::datatable((data.frame(reactome_gsea))[1:30,2:10]) -->
<!-- ``` -->

<!-- ## DO -->
<!-- ### Top 30 - Disease Ontology -->

<!-- ```{r print_do_do, echo=F} -->
<!-- DT::datatable((data.frame(disease_ontology_do))[1:30,2:10]) -->
<!-- ``` -->

<!-- ## NCG -->
<!-- ### Top 30 - Network of Cancer Gene -->

<!-- ```{r print_do_cng, echo=F} -->
<!-- DT::datatable((data.frame(disease_ontology_ncg))[1:30,2:10]) -->
<!-- ``` -->

<!-- ## DGN -->
<!-- ### Top 30 - Disease Gene Network -->

<!-- ```{r print_do_dgn, echo=F} -->
<!-- DT::datatable((data.frame(disease_ontology_dgn))[1:30,2:10]) -->
<!-- ``` -->

```{r print_degs, echo=F}
# Obtain HGNC symbols for DE genes
# degs_hgnc2 <-
#   biomaRt::getBM(
#     attributes = c('ensembl_gene_id', 'external_gene_name'),
#     filters = 'ensembl_gene_id',
#     values =  degs$Geneid,
#     mart = biomaRt::useDataset(
#       "mmusculus_gene_ensembl",
#       mart = biomaRt::useMart("ensembl", host = "useast.ensembl.org")
#     )
#   ) %>%
#   dplyr::rename('Geneid' = 'ensembl_gene_id',
#                 'hgnc' = 'external_gene_name') %>%
#   full_join(degs)

degs_hgnc2 <- degs %>% left_join(biomaRt_annotation_table %>% 
  dplyr::select(ensembl_gene_id, external_gene_name) %>% 
  dplyr::distinct(ensembl_gene_id, .keep_all = TRUE) %>% 
    dplyr::rename('Geneid' = 'ensembl_gene_id',
                  'hgnc' = 'external_gene_name'))

n_degs <- nrow(degs_hgnc2)

```


## Number of DEGs: `r n_degs`
```{r degs_table, echo =F}
#DT::datatable(degs_hgnc2 %>% arrange(padj))

#DT::datatable(degs)
n_degs <- nrow(degs)

```


<!-- Plot table -->
`r DT::datatable(degs_hgnc2[,1:8], rownames = FALSE, filter="top", options = list(pageLength = 10, scrollX=F, columnDefs = list(list(className = 'dt-center', targets = 0:(ncol(degs_hgnc2[,1:8])-1)))))`

## Heatmap

```{r degs_heatmap, echo = F, fig.height = 8, fig.width = 8.8}


degs_hgnc_heatmap <- degs_hgnc2 %>% 
  filter(!is.na(hgnc)) %>% 
  mutate(hgnc = str_to_title(hgnc))

normalized_counts_degs_heatmap <- normalized_counts_degs %>% 
  rownames_to_column(var = 'id')

# normalized_counts_degs_heatmap_hgnc <- biomaRt::getBM(
#     attributes = c('ensembl_gene_id', 'external_gene_name'),
#     filters = 'ensembl_gene_id',
#     values = normalized_counts_degs_heatmap$id,
#     mart = biomaRt::useDataset(
#         "mmusculus_gene_ensembl",
#         mart = biomaRt::useMart("ensembl", host = "ensembl.org")
#     )
# ) %>%
#     dplyr::rename('id' = 'ensembl_gene_id',
#                   'hgnc' = 'external_gene_name') %>%
#   full_join(normalized_counts_degs_heatmap)



normalized_counts_degs_heatmap_hgnc <- normalized_counts_degs_heatmap %>% left_join(biomaRt_annotation_table %>% 
  dplyr::select(ensembl_gene_id, external_gene_name) %>% 
  dplyr::distinct(ensembl_gene_id, .keep_all = TRUE) %>% 
    dplyr::rename('id' = 'ensembl_gene_id',
                  'hgnc' = 'external_gene_name'))

normalized_counts_degs_heatmap_hgnc_2 <- normalized_counts_degs_heatmap_hgnc %>% 
  filter(!is.na(hgnc)) %>% 
  distinct(hgnc, .keep_all = TRUE) %>% 
  column_to_rownames(var = 'hgnc') %>% 
  dplyr::select(-c('id'))

```

```{r interactive_DEGs, echo = FALSE}
# DEGs Heatmap ------------------------------------------------------------
# Construct the matrix with log2 transfored values to get the ~normalized distribution
normalized_counts_degs_heatmap <- normalized_counts_degs_heatmap_hgnc_2 %>%
  as.matrix() %>% 
  log2()
# Make the initial values equal 0 to be 0 again
normalized_counts_degs_heatmap[is.infinite(normalized_counts_degs_heatmap)] <- 0
# Save the colnames
column_names <- colnames(normalized_counts_degs_heatmap)
# Calculate the z-score for a better visualisation
normalized_counts_degs_heatmap_z <-
  base::apply(normalized_counts_degs_heatmap, 1, scale) %>%
  t() 
# Bring back colnames
colnames(normalized_counts_degs_heatmap_z) <- column_names
# Make NaN values 0 again
normalized_counts_degs_heatmap_z[is.nan(normalized_counts_degs_heatmap_z)] <- 0
# Paste the conditions
colnames(normalized_counts_degs_heatmap_z) <- paste(meta$condition, meta$samples, sep = "-")
# Order the heatmap according to conditions
normalized_counts_degs_heatmap_z_o <-
  normalized_counts_degs_heatmap_z[order(rownames(normalized_counts_degs_heatmap_z)), order(colnames(normalized_counts_degs_heatmap_z))]
# Create the gradient | heatmaply limits are broken so it will only take into the account limits as max and min matrix values | ComplexHeatmap can fix it but it is a shiny app that makes the markdown needed to be hosted
gradient_col <- ggplot2::scale_fill_gradient2(
  low = "blue", high = "red", 
  midpoint = 0
)
# Sort annotations
side_annot_top <- as.character(meta$condition) %>% 
  sort()
# Fix the height and width
degs_height = nrow(normalized_counts_degs_heatmap_z_o)*0.2
degs_width = ncol(normalized_counts_degs_heatmap_z_o)*0.32
```

### DEGs heatmap with ROW clustering

```{r heatmap_degs, echo = FALSE, fig.height = 10, fig.width = 7.5}
 
# Extract group names from comparison_name
groups_comparison <- strsplit(comparison_name, "_vs_")[[1]]
effect_group <- groups_comparison[1]
baseline_group <- groups_comparison[2]

 # Sort annotations
 side_annot_top <- meta$condition %>% 
   sort()

 col_fun2 = circlize::colorRamp2(c(-1.5, 0, 1.5),
                                 c('mediumblue', 'white', 'red2'), space = 'sRGB')


# colnames(normalized_counts_degs_heatmap_z_o) <- stringr::str_extract(colnames(normalized_counts_degs_heatmap_z_o), "[^_]*_[^_]*")


 ComplexHeatmap::Heatmap(normalized_counts_degs_heatmap_z_o, name = 'z-score', col = col_fun2,
                                  cluster_rows = T, cluster_columns = FALSE, show_row_names = F,
                                  border = F, show_column_names = TRUE,column_title = '',
                                  row_dend_width = unit(1, 'cm'), heatmap_legend_param = list(legend_height = unit(6, "cm")),
                        top_annotation = ComplexHeatmap::HeatmapAnnotation(Condition = side_annot_top, col = list(Condition = setNames(c("red", "blue"), c(effect_group, baseline_group))))
)
 degs_height_clustered = nrow(normalized_counts_degs_heatmap_z_o) * 0.25

```

### DEGs heatmap with ROW and COL clustering

```{r heatmap_degs_clustering, echo = FALSE, include = TRUE, fig.height = 10, fig.width = 7.5}

#colnames(normalized_counts_degs_heatmap_z_o) <- stringr::str_extract(colnames(normalized_counts_degs_heatmap_z_o), "[^_]*_[^_]*")


ComplexHeatmap::Heatmap(normalized_counts_degs_heatmap_z_o, name = 'z-score', col = col_fun2,
                                 cluster_rows = T, cluster_columns = T, show_row_names = F,
                                 border = F, show_column_names = TRUE,column_title = '',
                                 row_dend_width = unit(1, 'cm'), heatmap_legend_param = list(legend_height = unit(6, "cm")),
                       top_annotation = ComplexHeatmap::HeatmapAnnotation(Condition = side_annot_top, col = list(Condition = setNames(c("red", "blue"), c(effect_group, baseline_group))))
)


# sanity check
if (sum(metadata$sample == colnames(normalized_counts)) == length(metadata$sample)) {
  'complete'
} else {
  stop('metadata$sample == colnames(normalized_counts) is not met!')
}

```

